### ISSUE TITLE
`unnecessary_cast` false negative on noop calls to `ptr::cast`

### ISSUE THREAD

**Issue by lopopolo** [rust-lang/rust-clippy#8119]
### Summary

Calls to `<pointer>::cast` that do not change the type of the pointer contribute to code noise and are no-ops.

I'd like the complexity lint `unnecessary_cast` to warn on these.

### Lint Name

unnecessary_cast

### Reproducer

I tried this code:

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=1498de6e26fef7e004530e792de6b56e

```rust
#![warn(clippy::all)]
#![warn(clippy::pedantic)]

use std::ffi::CString;

extern "C" fn foo(ch: *const i8) {
    let _ = ch;
    todo!();
}

pub fn main() {
    let data = CString::new("foo").unwrap();
    foo(data.as_ptr().cast());
    foo(data.as_ptr().cast::<i8>());
    foo(data.as_ptr());
}
```

I expected to see this happen: calls to `cast` on the `CString` pointer should warn.

Instead, this happened: No warnings were emitted.

### Version

```text
Stable 1.57.0 from playground
```

**Comment by lopopolo** [comment]
I realize now that this is not a false negative because `CString`'s APIs take `c_char` which is a typedef with a `cfg`.
